pkg/reconcile/reconcile.go

```diff
@@ -110,6 +112,36 @@ var _ Reconciler = Func(nil)
 // Reconcile implements Reconciler.
 func (r Func) Reconcile(ctx context.Context, o Request) (Result, error) { return r(ctx, o) }
 
+// ObjectReconciler is a specialized version of Reconciler that acts on instances of client.Object. Each reconciliation
+// event gets the associated object from Kubernetes before passing it to Reconcile. An ObjectReconciler can be used in
+// Builder.Complete by calling AsReconciler. See Reconciler for more details.
+type ObjectReconciler[T client.Object] interface {
+	Reconcile(context.Context, T) (Result, error)
+}
+
+// AsReconciler creates a Reconciler based on the given ObjectReconciler.
+func AsReconciler[T client.Object](client client.Client, rec ObjectReconciler[T]) Reconciler {
```

Comment:
In https://github.com/kubernetes-sigs/controller-runtime/issues/2582#issuecomment-1819720649, @alvaroaleman proposed adding a new method to `builder.Builder` so that an adapter function like `AsReconciler` here wouldn't be necessary. However, Go does not allow a method to add new type parameters, only a method receiver can define type parameters. Therefore, I believe that adding a `Builder.CompleteObjectReconciler` method would only be possible if `Builder` defined a type parameter, i.e. `Builder[T client.Object]`. I don't think this type parameter makes much sense for instances of `reconcile.Reconciler` so I opted to use an adapter function instead.

Reply:
I actually realized the same thing later on while being on a walk, so agree 👍